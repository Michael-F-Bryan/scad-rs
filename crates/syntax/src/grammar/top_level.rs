use crate::{
    grammar::{expressions, TokenSet},
    parser::{Mark, Parser},
    SyntaxKind::*,
};

const CONTINUE: TokenSet = TokenSet::new([T![;], T!["}"]]);

pub(crate) fn package(p: &mut Parser<'_>) {
    let m = p.start();

    while !p.at(EOF) {
        statement(p);
    }

    p.complete(m, PACKAGE);
}

pub(crate) fn statement(p: &mut Parser<'_>) {
    let m = p.start();

    match p.current() {
        T![include] => include(p, m),
        T![use] => use_(p, m),
        IDENT if p.nth_at(1, T![=]) => {
            assignment(p, m);
            p.expect(T![;]);
        }
        _ => p.error_recover("Expected a statement", m, CONTINUE),
    }
}

pub(crate) fn include(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![include]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, INCLUDE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn use_(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![use]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, USE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn assignment(p: &mut Parser<'_>, m: Mark) {
    p.bump(IDENT);

    p.eat(T![=]);

    expressions::expr(p);
    p.complete(m, ASSIGNMENT);
}

#[cfg(test)]
mod tests {
    use super::*;

    parse_tests! {
        empty_package: package(""),
        include_statement: statement(r#"include "./foo/bar""#),
        use_statement: statement(r#"use "./foo/bar""#),
        assignment_statement: statement("x = 42;"),
        kitchen_sink: package(r#"
            include "foo/bar.scad"
            use "baz.scad"
            a = 42;
            b = (42);
            c = 1 + 1;
        "#),
    }
}
