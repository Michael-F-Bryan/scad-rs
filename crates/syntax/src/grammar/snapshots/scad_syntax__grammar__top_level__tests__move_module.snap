---
source: crates/syntax/src/grammar/top_level.rs
description: "statement(\n            module move(x=0,y) {\n                translate() rotate(5) children();\n                cube();\n            })"
---
Node(
    NAMED_MODULE_DEFINITION@0..57
      MODULE_KW@0..6 "module"
      IDENT@6..10 "move"
      L_PAREN@10..11 "("
      PARAMETERS@11..16
        ASSIGNMENT@11..14
          IDENT@11..12 "x"
          EQUALS@12..13 "="
          INTEGER@13..14
            INTEGER@13..14 "0"
        COMMA@14..15 ","
        IDENT@15..16 "y"
      R_PAREN@16..17 ")"
      L_CURLY@17..18 "{"
      MODULE_INSTANTIATION@18..49
        IDENT@18..27 "translate"
        L_PAREN@27..28 "("
        ARGUMENTS@28..28
        R_PAREN@28..29 ")"
        MODULE_INSTANTIATION@29..49
          IDENT@29..35 "rotate"
          L_PAREN@35..36 "("
          ARGUMENTS@36..37
            INTEGER@36..37
              INTEGER@36..37 "5"
          R_PAREN@37..38 ")"
          MODULE_INSTANTIATION@38..49
            IDENT@38..46 "children"
            L_PAREN@46..47 "("
            ARGUMENTS@47..47
            R_PAREN@47..48 ")"
            SEMICOLON@48..49
              SEMICOLON@48..49 ";"
      MODULE_INSTANTIATION@49..56
        IDENT@49..53 "cube"
        L_PAREN@53..54 "("
        ARGUMENTS@54..54
        R_PAREN@54..55 ")"
        SEMICOLON@55..56
          SEMICOLON@55..56 ";"
      R_CURLY@56..57 "}"
    ,
)
